examples/attestation/main.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2023 Mariano Cano
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"log"
+	"math/big"
+	"net/http"
+	"os"
+
+	"github.com/mholt/acmez"
+	"github.com/mholt/acmez/acme"
+	"go.uber.org/zap"
+)
+
+const usage = `Usage: STTY=-icanon attestation <csr-file>
+
+<csr-file> A file with a certificate signing request or CSR.
+
+To be able to run this example, we need to use a key that can be attested,
+"step-ca" [1], for example, supports attestation using YubiKey 5 Series.
+
+To configure "step-ca" with device-attest-01 support, you need to create an ACME
+provisioner with the device-attest-01 challenge enabled. In the ca.json the
+provisioner looks like this:
+
+  {
+    "type": "ACME",
+    "name": "attestation",
+    "challenges": [ "device-attest-01" ]
+  }
+
+After configuring "step-ca" the first thing that we need is to create a key in
+one of the YubiKey slots. We're picking 82 in this example. To do this, we will
+use "step" [2] with the "step-kms-plugin" [2], and we will run the following:
+
+  step kms create "yubikey:slot-id=82?pin-value=123456"
+
+Then we need to create a CSR signed by this new key. This CSR must include the
+serial number in the Permanent Identifier Subject Alternative Name extension.
+The serial number of a YubiKey is printed on the key, but it is also available
+in an attestation certificate. You can see it running:
+
+  step kms attest "yubikey:slot-id=82?pin-value=123456" | \
+  step certificate inspect
+
+To add the permanent identifier, we will need to use the following template:
+
+  {
+    "subject": {{ toJson .Subject }},
+    "sans": [{
+      "type": "permanentIdentifier",
+      "value": {{ toJson .Subject}}
+    }]
+  }
+
+Having the template in "attestation.tpl", and assuming the serial number is
+123456789, we can get the proper CSR running:
+
+  step certificate create --csr --template attestation.tpl 123456789 att.csr
```

Comment:
I think this needs the `yubikey://` to be provided as the private key?

Reply:
Fixed with d30f7c5